### DISCUSSION TITLE
Unwanted system libraries: /lib/x86_64-linux-gnu/libgmp.so.10

### DISCUSSION THREAD

**Discussion by aryairani** [Homebrew/discussions#1638] [Tap maintenance and Homebrew development] [answered]
```
==> brew linkage --test aryairani/unison/ucm
==> FAILED
Unwanted system libraries:
  /lib/x86_64-linux-gnu/libgmp.so.10
```

It looks like `gmp` is a transitive dependency of my `:build` dependency in linux ([haskell-stack.rb](https://github.com/Homebrew/linuxbrew-core/blob/master/Formula/haskell-stack.rb#L27-L29)). I don't get why `:build` dependencies would be sticking around, and I don't see that I have any control over how to solve this.

Any suggestions, or clarifying questions I can answer?  Thanks in advance.

My full formula: [ucm.rb](https://github.com/aryairani/homebrew-unison/blob/abc87959daea78d1c1e9dc416fb0de825b5c2120/Formula/ucm.rb)

My full test-bot log: https://github.com/aryairani/homebrew-unison/runs/2761066014

**Comment by carlocab** [comment, marked as answer]
`/lib/x86_64-linux-gnu/libgmp.so.10` does not come from the `gmp` formula (the `gmp` formula should have a path that contains `linuxbrew/.linuxbrew`), but is a system-provided library that's available on the system you happen to be building on.

You can probably avoid this linkage by using a Homebrew `ghc` to build, rather than letting `stack` download a `ghc` to use for the build. Letting `stack` do that gives you no control over what libraries the build-time `ghc` will use.

**Reply by aryairani** [reply]
Oh, hm, I see. What is `brew linkage --test aryairani/unison/ucm` actually saying? That the executable I've built in `def install` references that system library (which we're not sure where it came from) but shouldn't?

Letting `stack` download a `ghc` for the build shouldn't be adding anything to `/lib`; that's not even normally even writable, right? Is there some way to track down when that addition actually occurs?

**Reply by carlocab** [reply]
Nothing is being added to `/lib`. The problem is that the system you're building on has a `libgmp` there, so the `ghc` that `stack` downloaded is linking with it opportunistically (perhaps `stack` is even downloading a `ghc` that uses a `libgmp` that lives in `/lib` because it knows the system you're building on has the appropriate library there).

This means that whatever you've built is now no longer portable, as it can only be installed into systems with `/lib/x86_64-linux-gnu/libgmp.so.10`.

**Reply by aryairani** [reply]
I see, it's nice that brew checks this.  Thanks!

**Reply by danielnachun** [reply]
Right now I think we build all our Haskell packages using `cabal-install`.  As long as `unison` works on a version of `ghc` we support, I'd like to think it's possible to build `unison` that way instead of using `stack`.  Though my knowledge of Haskell packaging is limited!

**Reply by carlocab** [reply]
We do build a handful of packages using `stack`, though I believe all of them tell `stack` to use a Homebrew GHC for the build.

**Reply by danielnachun** [reply]
That's great that `stack` allows specifying the Homebrew GHC instead.  In that case I guess the `stack install` call should be tweaked, and `ghc` added as dependency.

**Reply by aryairani** [reply]
@danielnachun We like that `stack` will auto-install a tested, compatible version of `ghc` for unison contributors building from source (and CI) instead of manually trying to ensure that the ghc version matches, and we're trying to avoid having to support both `cabal-install` and `stack` project definitions. 😓  

It's a bummer that the brew-installed `stack` doesn't produce brew-friendly executables, but I can imagine reasons why it might be hard to make that happen.

**Reply by carlocab** [reply]
It's not just that the executables are brew-friendly: the executables you build with `stack` are not portable. I'm fairly certain (though would be pleased to be proved wrong) that even a non-brew-installed `stack` does this.

As I've already mentioned, you can get around this by getting `stack` to use a Homebrew GHC, rather than pulling whatever GHC it wants to out of its resolver. If you'd rather not, that's fine too, but you do pay the price of the executable you build not being portable.

If you want to keep using `stack` to pull in whatever GHC it wants, you can try doing `depends_on "gmp"` to see if it'll link with Homebrew GMP instead (because that's something you can guarantee will exist on any platform you wish to install your formula in). You might need some extra flags to persuade your executable to do this, though.

**Reply by danielnachun** [reply]
If it's possible to supply the equivalent of `-L$HOMEBREW_PREFIX/opt/gmp/lib` and `-rpath=$HOMEBREW_PREFIX/opt/gmp/lib` when building GHC with Stack, then that should make it link to brewed `gmp` instead of the system one.

Out of curiosity, does `unison` need a specific minor version of `ghc`, or is it just the major version that needs to be controlled?  We do support all major versions of GHC which are still supported by upstream as separate formulae (which currently include 8.6, 8.8, 8.10 and 9.0), so it is possible to control the major version at least.  I would hope that GHC follows the usual major/minor release model where the minor versions are ABI/API compatible and the focus is mostly on bug fixes and performance improvements, but I'm not sure.  

If you really do need to control the minor version, you could still use brewed GHC.  You can use `brew extract` to get a specific minor version of GHC and add it to your tap.  This could probably be automated so that it taps whatever version is used for the Stack config whenever that file is updated.  

If a project depends on an upstream source which doesn't maintain API/ABI compatibility across minor versions this is a tough situation for package managers and developers.  As package maintainers we have no desire or bandwidth to maintain minor versions of a dependency that upstream is not maintaining any more (i.e. GHC 8.10.1 is probably considered unsupported as soon as GHC 8.10.2 is released).  As an upstream developer I'm sure you don't want to have to fix a bunch of breakages after every minor version update.  That may not be what's happening here, but it has been a problem for other projects and it can be a huge headache.

**Reply by aryairani** [reply]
@carlocab 

> It's not just that the executables are brew-friendly: the executables you build with `stack` are not portable.

Right — I just meant it would be convenient if the executables built with the stack that comes from the brew formula would be portable as far as the homebrew ecosystem is concerned; maybe that's not even a wish that makes sense, though.

e.g. if the GHCs it installed were somehow already linked to `$HOMEBREW_PREFIX/opt/gmp/lib`.

> I'm fairly certain (though would be pleased to be proved wrong) that even a non-brew-installed `stack` does this.

I would guess so too, but with less expectation that it would do anything better.

> As I've already mentioned, you can get around this by getting `stack` to use a Homebrew GHC, rather than pulling whatever GHC it wants to out of its resolver. If you'd rather not, that's fine too, but you do pay the price of the executable you build not being portable.

Yes, I have made this change currently — it's okay in brew formula, but a little harder in the context of `cabal-install`, because here I have to remember to do it, and there someone else has to remember to ;-)

> If you want to keep using `stack` to pull in whatever GHC it wants, you can try doing `depends_on "gmp"` to see if it'll link with Homebrew GMP instead (because that's something you can guarantee will exist on any platform you wish to install your formula in). You might need some extra flags to persuade your executable to do this, though.

Good idea, I might try this too; thanks!

**Reply by aryairani** [reply]
@danielnachun 

> If it's possible to supply the equivalent of `-L$HOMEBREW_PREFIX/opt/gmp/lib` and `-rpath=$HOMEBREW_PREFIX/opt/gmp/lib` when building GHC with Stack, then that should make it link to brewed `gmp` instead of the system one.

Interesting, I'll look into this.

> Out of curiosity, does `unison` need a specific minor version of `ghc`

No, it shouldn't; the only issue we have seen with minor versions was that <8.10.4, Big Sur wasn't supported.

Thanks for your help!

**Reply by danielnachun** [reply]
If `unison` doesn't depend on minor versions of GHC and meets the requirements for submission to homebrew-core, I think the easiest way to get `unison` to users is by just opening a PR for a new formula in homebrew-core, using `stack` to build it with brewed GHC.  

Right now the default `ghc` formula is at 8.10 - that will eventually change to 9. When it does, our CI will try to build everything that depends on `ghc` with GHC 9.  If `unison` failed to build because it did not yet support GHC 9, we would make it depend on `ghc@8.10`.  Once the `unison` upstream tagged a release that officially supported GHC 9, we would change the dependency back to `ghc`.  As long as `unison` is actively maintained so that it will eventually support new major versions of GHC, our maintainers can handle making it depend on an older major version of GHC while the upstream team works on supporting the newest major version.  This is a routine part of our normal maintenance work that happens when compilers and other core dependencies are updated.
